src/code_190B20.c

```diff
@@ -284,23 +281,19 @@ INCLUDE_ASM(s32, "code_190B20", func_8026709C);
 
 INCLUDE_ASM(s32, "code_190B20", func_802670C8);
 
-#ifdef NON_MATCHING
-// Register allocation issues (decorationIndex is placed in s2 for seemingly no reason?).
-// Should be easy to clean up once DecorationTable is more understood
 void add_part_decoration(ActorPart* part, s32 decorationIndex, DecorationId decorationType) {
-    if (part->idleAnimations && (part->flags & 2) == 0) {
-        DecorationTable* decoration = &part->decorationTable->unk_00[decorationIndex];
+    if ((part->idleAnimations) && !(part->flags & 2)) {
+        char* decorationTable = part->decorationTable->unk_00;
+        DecorationTable* decoration;
+
         _remove_part_decoration(part, decorationIndex);
+        decoration = decorationTable + decorationIndex;
```

Comment:
It'd be nice if there's some way to do this by using proper array index accessing syntax

Reply:
Agreed; it looks like this is what I tried to do originally, which made this not match. Perhaps we'll work it out later when we understand the decorationTable better.

Reply:
Agreed; it looks like this is what I tried to do originally, which made this not match. Perhaps we'll work it out later when we understand the decorationTable better.

Reply:
This, in particular, seems just wrong. Likely, there's a bunch of arrays in the struct (ones which we haven't defined yet) and those arrays are most likely all some byte-sized arrays. So the compiler can optimize accessing each array with this index into just adding the index to the base offset of the struct and then accessing everything else from a fixed offset from there. I think this should most likely be explored before we just merge this.

Reply:
Yeah, my above suspicions were correct. I modified the struct a bit, though it may still not be correct, and managed to get a match: https://hastebin.com/ogocunazos.cpp

Still not sure if this struct is correct, but the optimization the compiler does is confirmed with this. Not sure how I feel about having this struct and func in here, since we're still not sure that it's correct.